### PR TITLE
Fix FocusChanged routing

### PR DESCRIPTION
**Pull request by cmyr**
~This is based off of #499, which should go in first.~

This fixes *two* issues in focus routing; the first is that we weren't correctly checking children when deciding to recurse, and the second is that we were converting from `RouteFocusChanged` to `FocusChanged` without correctly checking if the current widget was actually involved in the change. 😬

### CODE REVIEW

File: druid/src/core.rs [hunk 1]
```diff
@@ -423,15 +423,25 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
             LifeCycle::HotChanged(_) => false,
             LifeCycle::RouteFocusChanged { old, new } => {
                 self.state.request_focus = None;
-                let this_changed = old.map(|_| false).or_else(|| new.map(|_| true));
+
+                let this_changed = if *old == Some(self.state.id) {
```
Comment (Zarenor) [inline druid/src/core.rs]: I think this respelling is much better. It took me several minutes tease apart the old version the first time.

File: druid/src/core.rs [hunk 2]
```diff
@@ -423,15 +423,25 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
                     let event = LifeCycle::FocusChanged(change);
                     self.inner.lifecycle(ctx, &event, data, env);
                     false
                 } else {
-                    //FIXME: I think this is a bug? we should be checking `self.state.children`
-                    old.map(|id| ctx.children.contains(&id)).unwrap_or(false)
-                        || new.map(|id| ctx.children.contains(&id)).unwrap_or(false)
+                    old.map(|id| self.state.children.contains(&id))
+                        .unwrap_or(false)
+                        || new
```
Comment (Zarenor) [inline druid/src/core.rs]: Minor style nit: The or operator here tripped me up ("Why is there a closure here?!")
Can we spell this
`[...]
.unwrap_or(false)
||
new.map(...)
[...]`
Or will rustfmt move it back this way?

Comment (cmyr) [reply druid/src/core.rs]: ah interesting, I hadn't thought to see that as a closure. This formatting is rustfmt's doing, but I can rewrite this using `or_else`, which should be clearer.

Comment (Zarenor) [reply druid/src/core.rs]: Thanks. I only just now realized that GH discarded my newlines in what I'd written. Edited above to what I intended.

**Pull request merged**